cobalt/browser/cobalt_content_browser_client.cc

```diff
@@ -295,4 +327,118 @@ bool CobaltContentBrowserClient::WillCreateURLLoaderFactory(
+  }
+}
+
+void CobaltContentBrowserClient::CreateFeatureListAndFieldTrials() {
+  CreateExperimentConfig();
+  SetUpFieldTrials();
+}
+
+void CobaltContentBrowserClient::SetUpFieldTrials() {
+  metrics::TestEnabledStateProvider enabled_state_provider(/*consent=*/false,
+                                                           /*enabled=*/false);
+  base::FilePath path;
+  base::PathService::Get(content::SHELL_DIR_USER_DATA, &path);
+  std::unique_ptr<metrics::MetricsStateManager> metrics_state_manager =
+      metrics::MetricsStateManager::Create(
+          exp_config_.get(), &enabled_state_provider, std::wstring(), path,
+          metrics::StartupVisibility::kForeground);
+  metrics_state_manager->InstantiateFieldTrialList();
+
+  // std::vector<std::string> variation_ids;
+  auto feature_list = std::make_unique<base::FeatureList>();
+
+  const base::CommandLine& command_line =
+      *base::CommandLine::ForCurrentProcess();
+
+  // Overrides for content/common and lower layers' switches.
+  std::vector<base::FeatureList::FeatureOverrideInfo> feature_overrides =
+      content::GetSwitchDependentFeatureOverrides(command_line);
+
+  // Overrides for content/shell switches.
+
+  // Overrides for --run-web-tests.
+  if (switches::IsRunWebTestsSwitchPresent()) {
+    // Disable artificial timeouts for PNA-only preflights in warning-only mode
+    // for web tests. We do not exercise this behavior with web tests as it is
+    // intended to be a temporary rollout stage, and the short timeout causes
+    // flakiness when the test server takes just a tad too long to respond.
+    feature_overrides.emplace_back(
+        std::cref(
+            network::features::kPrivateNetworkAccessPreflightShortTimeout),
+        base::FeatureList::OVERRIDE_DISABLE_FEATURE);
+  }
+
+  feature_list->InitializeFromCommandLine(
+      command_line.GetSwitchValueASCII(::switches::kEnableFeatures),
+      command_line.GetSwitchValueASCII(::switches::kDisableFeatures));
+
+  // This needs to happen here: After the InitFromCommandLine() call,
+  // because the explicit cmdline --disable-features and --enable-features
+  // should take precedence over these extra overrides. Before the call to
+  // SetInstance(), because overrides cannot be registered after the FeatureList
+  // instance is set.
+  feature_list->RegisterExtraFeatureOverrides(feature_overrides);
+
+  // Add Cobalt custom code to create field trial here.
+  base::FieldTrial* cobalt_field_trial = base::FieldTrialList::CreateFieldTrial(
+      kCobaltExperimentName, kCobaltGroupName);
+  CHECK(cobalt_field_trial) << "Unexpected name conflict.";
+
+  const base::Value::Dict& feature_map =
+      exp_config_->GetDict(kExperimentConfigFeature);
+  const base::Value::Dict& param_map =
+      exp_config_->GetDict(kExperimentConfigFeatureParams);
+
+  for (const auto kvpair : feature_map) {
+    if (kvpair.second.is_bool() && kvpair.second.GetBool()) {
```

Comment:
We should through a fatal error if they're not boolean, ya?

Reply:
Is it necessary for fatal error? If not overridden, it'll just use the default value which should be safe. We could log an error message.

Reply:
I think I prefer fatal as it really is a fundamentally bad state to be in. The client should not be passing parameters of types that are unexpected. If they do, that's broken. @jellefoks may have insight into what deserves to be a FATAL vs. not.

Reply:
Per our offline conversation, logging an error and monitoring it offline seems like a reasonable way to go. This avoids unexpected outages and also allows us to catch issues. The key is going to be returning a rejected promise from the h5vcc API and also logging a UMA histogram here. I've updated https://b.corp.google.com/issues/407734134 to also mention the monitoring piece. I'd say feel free to carry on with this PR as-is.